requests-for-comments/2021-10-05-group-analytics.md

```diff
@@ -0,0 +1,214 @@
+- Feature flags: roll out feature group-by-group
+
+For PostHog the product, the 3 group types we would use are:
+
+- Instances (e.g. cloud, posthog.someproduct.org, …)
+- Organizations
+- Projects
+
+## Semantics & Constraints
+
+### 1. Users can belong to multiple groups
+
+Some examples:
+
+1. Contractor type work (e.g. for farm management agronomists work for multiple farms at once)
+2. Billing happens on a different level (e.g. our projects vs organization)
+3. Users clients are global companies with a lot of local “branches” where execs need access to everything
+4. Internal users have access to debug issues on multiple accounts
+
+### 2. User is active in only “group” at any point in time
+
+Whenever an event is made, we can (and should) make the assumption that it belongs to a single group (e.g. organization). This is needed to avoid “double-counting” in stats.
+
+Example: If Yakko has been added to 20 different organizations and is active, we don’t want to count all of them as active.
+
+### 3. Limited amount of group types
+
+From speaking with clients, their internal amount of different “group types” (e.g. `organization`) is limited, though the number of groups under a group type is unlimited (e.g. Facebook if it was an organization).
+
+This constraint allows us to implement things slightly more efficiently and have a nicer UX compared to unbounded amount of groups.
+
+### 4. Feature flags in groups should be consistent
+
+For the users who are using feature flags, they want to ensure a consistent experience within a group during rollout.
+
+This is to avoid situations where some users see a new feature and some do not.
+
+This also factors into their later analysis - for example, they’d like to treat two different “organizations” both as a single datapoint even if one has 1500 users and the other has 1.
+
+## Proposal
+
+### Groups are event-level
+
+Users can define up to N (5?) group types. For us these would be `instance` , `project` and `organization`.
+
+Each event we ingest can have one group per group type associated.
+
+We will have N new columns in the `events` table, containing the group ids - first one for the first group type of the project, second for the second one etc.
+
+### Group properties API
+
+We will expose a new API for creating and adding properties groups. This API can be exposed on the client level, e.g.
+
+```python
+    import posthog
+
+    posthog.groups.add_properties('organization', 'org:5', { name: 'Swedish Bakery', subscription_level: 'gold' })
+```
+
+There will not be any functionality for merging two groups, as that behavior is user-specific.
+
+### Capture API
+
+Capturing events in client libraries will allow users to pass group metadata alongside with an event.
+
+Here’s how it would look like in python:
+
+```python
+    import posthog
+
+    posthog.capture(
+      user_id,
+      'purchase'
+      { plan: 'silver' },
+      groups={
+        'organization': 'org:6'
```

Comment:
+1 to above. Also, in general unclear on how the group information is stored for persons

Reply:
> What do we do when a new group type appears? 

Should we put this in the RFC?

> If no such column found but there are free columns for that project, we put it in the next one

Typos in a code can create columns, which means the user can be left with odd testing names for the columns. So we should add a way to be able to rename the columns. Doesn't need to be in the MVP. 

> If out of columns, we ignore

This will likely come to as a surprise to users at one point & cause confusion. Is there a way to assign a group later? If we had `group_assignment_error` or something property added to the event, we could at least discover the problem. Doesn't need to be in the MVP. But maybe add "TBD reporting the failure"